### PR TITLE
RF extract complex methods out from signal.clean 

### LINKED ISSUE #2820: Refactor `signals.clean`
The current code in `signals.py`, especially `signal.clean` is difficult to follow. To facilitate the integration of `load_confounds` #2777, and easier integration of DCT #1011, we should refactor it to a level where the data flow is easier to follow.

I would suggest, for now, extract some methods out just to clean up `signal.clean`. We can discuss if this function should be written into a module.

### PR DESCRIPTION
**Pull request by htwangtw**
Closes #2820 
The aim of this PR is to simplify the data flow inside `signal.clean` for future maintenance.
I have left a few place holder for DCT #1011 @jerdra

I added an input argument `filter` for specifying the filtering method, preempting DCT to be added. Currently, I set the default to "Butterworth" to prevent breaking the tests. If we want that part to behave more logically, further work will be needed.

### CODE REVIEW

File: nilearn/signal.py
Comment (bthirion) [inline nilearn/signal.py:20]:
```suggestion
available_filters = ['butterworth', ]
```

File: nilearn/signal.py
Comment (bthirion) [inline nilearn/signal.py:413]:
```suggestion
          confounds=None, standardize_confounds=True, filter='butterworth',
```

File: nilearn/signal.py
Comment (bthirion) [inline nilearn/signal.py:571]: I think that what is called a session here is actually a run in BIDS language.

File: nilearn/signal.py
Comment (bthirion) [inline nilearn/signal.py:512]:
```suggestion
    # check if filter parameters are satisfied
```

File: nilearn/signal.py
Comment (bthirion) [reply nilearn/signal.py:571]: Oh, what you mean is that you are processing a set of runs, so rather `_process_runs` maybe

File: nilearn/signal.py
Comment (bthirion) [inline nilearn/signal.py:412]: maybe we should deprecate `sessions` here, so that we can use  `runs` instead in a future release (0.9).

**Review by bthirion — COMMENTED** [review]
Thx ! looks quite good already.
There a re a few lines not covered by tests.

File: nilearn/signal.py
Comment (htwangtw) [reply nilearn/signal.py:571]: I agree that's a more sensible name

File: nilearn/signal.py
Comment (htwangtw) [reply nilearn/signal.py:412]: Sounds good. May I ask if there's an example for implementing deprecation warning in nilearn? I have never handled this kind of things before.

**Review by bthirion — COMMENTED** [review]
Thx, this is looking pretty good.

**Comment by htwangtw** [comment]
I tried to add the cosine drift method from design_matrix but that introduced an error while running the test:

```
ImportError while loading conftest '/home/hw1012/git_repos/nilearn/nilearn/conftest.py'.
nilearn/conftest.py:9: in <module>
    from nilearn.datasets._testing import request_mocker  # noqa: F401
nilearn/datasets/__init__.py:6: in <module>
    from .struct import (fetch_icbm152_2009, load_mni152_template,
nilearn/datasets/struct.py:16: in <module>
    from ..image import new_img_like, get_data
nilearn/image/__init__.py:5: in <module>
    from .resampling import resample_img, resample_to_img, reorder_img, \
nilearn/image/resampling.py:16: in <module>
    from .image import crop_img
nilearn/image/image.py:19: in <module>
    from .. import signal
nilearn/signal.py:17: in <module>
    from .glm.first_level.design_matrix import _cosine_drift
nilearn/glm/__init__.py:8: in <module>
    from nilearn.glm.contrasts import (
nilearn/glm/contrasts.py:14: in <module>
    from nilearn.input_data import NiftiMasker
nilearn/input_data/__init__.py:5: in <module>
    from .nifti_masker import NiftiMasker
nilearn/input_data/nifti_masker.py:12: in <module>
    from .base_masker import BaseMasker, filter_and_extract
nilearn/input_data/base_masker.py:15: in <module>
    from .. import masking
nilearn/masking.py:15: in <module>
    from .image import new_img_like
E   ImportError: cannot import name 'new_img_like' from 'nilearn.image' (/home/hw1012/git_repos/nilearn/nilearn/image/__init__.py)
```

I am certain it is related to `from .glm.first_level.design_matrix import _cosine_drift`, but I am not sure why or how?
I can either hold off adding `filter='cosine'` in the current code or resolve this import error. 
What do you think?

**Comment by NicolasGensollen** [comment]
Hi @htwangtw 
You're getting a circular import issue. I think one simple solution might be to put the import (`from .glm.first_level.design_matrix import _cosine_drift`) inside the function which needs it (`_filter_signal`).
Let me know if this solves the issue!

**Comment by jerdra** [comment]
Two quick questions related to this PR:

- Is there a reason why the `_cosine_drift` function could not be extended to perform low-pass? 
- Would it make sense to also add an option for fourier drift? I know that some people prefer this (although I'm aware that they yield similar results in the end)
 
P.S: Let me know if the issues are a more appropriate place to discuss this!

**Comment by htwangtw** [comment]
I am think perhaps I will finish the refactoring, and implement the existing dct function. Once merged, @jerdra can open a separate issue to improve dct and add other filtering functions.
What do you think?

**Comment by NicolasGensollen** [comment]
@htwangtw @jerdra I agree, I think it would be easier to keep this PR on the pure refactoring of signal.clean and have other discussions within dedicated issues/PRs.

File: nilearn/signal.py
Comment (htwangtw) [reply nilearn/signal.py:581]: The last column of output from _cosine_drift will always be a column of ones, but it doesn't hurt to do a check here.

Comment (jerdra) [reply nilearn/signal.py:581]: Would there be an issue with the bias term of `_cosine_drift` being redundant with `standardize=True`? Since the mean will be removed anyway? Or is this a non-problem?

**Comment by htwangtw** [comment]
So the cosine drift filter is currently just added the existing function from the `design_matrix` module. Currently, I only added tests to check the outputs of private function `_filter_signal` with `filter='cosine'` are the expected shape/values. Would it be okay to leave the proper implementation of DCT to another PR, and a proper test on the cleaned signal?

File: nilearn/signal.py
Comment (bthirion) [inline nilearn/signal.py:445]: Here we are changing the API. This needs a deprecation cycle (raise a warning, saying that sessions will result in an error after release 0.9)

File: nilearn/signal.py
Comment (bthirion) [inline nilearn/signal.py:595]:
```suggestion
    for run in np.unique(runs):
```(nitpick, but this is more readable)

File: nilearn/tests/test_signal.py
Comment (bthirion) [inline nilearn/tests/test_signal.py:494]: Maybe indicate that this is supposed t fail due to t_r being None

File: nilearn/tests/test_signal.py
Comment (bthirion) [inline nilearn/tests/test_signal.py:713]: There may be an indentation problem here.

**Review by bthirion — COMMENTED** [review]
This is really looking great, thx !
Please also add a whatsnew entry.

File: nilearn/signal.py
Comment (htwangtw) [reply nilearn/signal.py:445]: Yes this has been done in L414 with `rename_parameters`! I will add the description to the docstring as well.

File: nilearn/signal.py
Comment (bthirion) [inline nilearn/signal.py:590]: Leave only 2 blank lines.

File: nilearn/signal.py
Comment (bthirion) [inline nilearn/signal.py:652]: It is good to give the actual and expected value in the error message.

**Review by bthirion — COMMENTED** [review]
This is great, thx !
Just nitpicking on 2 details, but otherwise it's ready to go.

**Review by bthirion — APPROVED** [review]
AFAIK the circleci is unrelated.  
Thx !

**Comment by NicolasGensollen** [comment]
@htwangtw I pushed a temporary fix on master which enables us to build the doc while the issue is resolved on sphinx's side. 
Could you merge master to fix the CI on this branch?
Thanks!

**Review by NicolasGensollen — APPROVED** [review]
LGTM, thanks @htwangtw !

**Comment by bthirion** [comment]
Merging, thx !

**Pull request merged**
